### PR TITLE
Implement dss logs functionality.

### PR DESCRIPTION
**Pull request by misohu**
Please first review this one https://github.com/canonical/data-science-stack/pull/62

This PR resolves issue: https://github.com/canonical/data-science-stack/issues/33

1. Implemented `dss logs` with `--mlflow` and `--all` flags
2. Implemented unit test for happy and bad scenarios 
3. Implemented integration test

How to test: 
If you want to test locally run the steps from .github/workflows/tests.yaml:integration (integration tests)

### CODE REVIEW

File: setup.cfg [hunk 1]
```diff
@@ -13,7 +13,7 @@ install_requires =
 include_package_data = True
 
 [options.package_data]
-dss = manifests.yaml
+dss = manifest_templates/*
```
Comment (DnPlas) [inline setup.cfg:16]: This definitely works, but it sounds like it'd be better in its own PR. You can also mention it is part of #44.

File: src/dss/logs.py [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+            return
+    except ApiError:
+        deployment = None
+
+    # The get command can return empty DeploymentList object with empty spec field
+    if deployment is None or deployment.spec is None:
+        if parts == "notebooks":
+            logger.error(f"No notebook found with the name '{name}'.")
+            logger.info(
+                "Please wait a moment or recreate the notebook with 'dss create-notebook'."
```
Comment (DnPlas) [inline src/dss/logs.py:42]: Ideally we'd have a command that would list the notebooks and show their status(?), that way in this error msg we can have something like "Verify the notebook is running with <dss command> or recreate the notebook...". It doesn't have to be part of this PR, but something that can provide a better UX. What do you think?

File: src/dss/logs.py [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+            logger.error(f"No pods found for notebook - {name}.")
+            logger.info(
+                "Please wait a moment or recreate the notebook with 'dss create-notebook'."
+            )
+        elif parts == "mlflow":
+            logger.error("No pods found for MLflow.")
+            logger.info("Please wait a moment or run 'dss initialize' first.")
+        return
+
+    for pod in pods:
```
Comment (DnPlas) [inline src/dss/logs.py:63]: Please correct me if I am wrong. I understand this will print the logs from the Notebook server and from each of the Jupyter notebooks that get created via `dss create-notebook` OR from the web app, right?

If that's the case, I think the output will be huge if users have a lot of notebooks. Is there a way to filter out a bit more the result? Say, by specifying the name of the actual notebook (not the notebook server)?

File: src/dss/main.py [hunk 4]
```diff
@@ -30,5 +31,37 @@ def initialize_command(kubeconfig: str) -> None:
     initialize(lightkube_client=lightkube_client)
 
 
+@main.command(name="logs")
+@click.option(
+    "--parts",
+    type=click.Choice(["notebooks", "mlflow"]),
+    required=True,
+    help="Specify which part's logs you want to retrieve: 'notebooks' or 'mlflow'.",
```
Comment (DnPlas) [inline src/dss/main.py:39]: I was having a hard time understanding what `part` means. Would it make sense to have something like `dss logs <mlflow|notebooks> <--name=NAME>`? And name can be an option exclusively for Notebooks so you don't have to do the `Warning: The 'name' parameter will be ignored for 'mlflow' logs.` workaround.

**Review by DnPlas — COMMENTED** [review]
Thanks @misohu, I have left some comments mainly focusing on the UX.

**Comment by misohu** [comment]
This PR needs to be postponed until this is resolved https://github.com/canonical/data-science-stack/issues/58

**Comment by misohu** [comment]
The dss remove command has been discussed with UX team please check the DSS spec [here ](https://docs.google.com/document/d/1v0wU8TBb-ui6RcjGRbLwA6XFZnWbuiVb2TPmEgwKejY/edit#heading=h.2881deebboks) for details about the error messages and parameters.

File: src/dss/logs.py (hunk 3 above)
Comment (misohu) [reply src/dss/logs.py:81]: When we will run 
```
dss create notebook-name
```
that command will create exactly one Deployment with one Pod and inside will be one container with jupyter-lab instance (lets call it lab because server is confusing). Inside the lab users will create notebooks but everything will stay in one pod in one container. There is no simple way to extract logs only for one notebook. From my practical experience you want to see the labs logs if there is some kernel problem or web app problem. I cannot even come up with some practical logs which would be specific only to the notebooks.

File: src/dss/logs.py [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+from lightkube.resources.core_v1 import Pod
+
+from dss.config import DSS_NAMESPACE, MLFLOW_DEPLOYMENT_NAME
+from dss.logger import setup_logger
+
+# Set up logger
+logger = setup_logger("logs/dss.log")
+
+
+def get_logs(parts: str, name: str, lightkube_client: Client) -> None:
```
Comment (orfeas-k) [inline src/dss/logs.py:13]: What about keeping the argument as `notebook_name`, since that's what the caller function uses and it makes for a more transparent name.

File: src/dss/main.py [hunk 6]
```diff
@@ -30,5 +31,43 @@ def initialize_command(kubeconfig: str) -> None:
...
+        get_logs("mlflow", None, lightkube_client)
+    elif notebook_name:
+        get_logs("notebooks", notebook_name, lightkube_client)
+
+
+# Define Examples section for logs command
+logs_command.help += """
+\n\nExamples:\n
+  dss logs my-notebook\n
+  dss logs --mlflow
```
Comment (orfeas-k) [inline src/dss/main.py:69]: We could include `--all` here as well.

File: src/dss/main.py [hunk 7]
```diff
@@ -30,5 +31,43 @@ def initialize_command(kubeconfig: str) -> None:
...
+    if print_all:
+        get_logs("all", None, lightkube_client)
+    elif mlflow:
+        get_logs("mlflow", None, lightkube_client)
+    elif notebook_name:
+        get_logs("notebooks", notebook_name, lightkube_client)
+
+
+# Define Examples section for logs command
+logs_command.help += """
```
Comment (orfeas-k) [inline src/dss/main.py:66]: Is this how a coomand `help` field should be defined? I mean outside its `def`inition? It's more of a question since it feels a bit weird but I 'm not a Python expert either.

File: src/dss/main.py [hunk 8]
```diff
@@ -30,5 +31,43 @@ def initialize_command(kubeconfig: str) -> None:
...
+)
+@click.argument("notebook_name", required=False)
+@click.option(
+    "--all", "print_all", is_flag=True, help="Print the logs for all notebooks and MLflow."
+)  # noqa: E501
+@click.option("--mlflow", is_flag=True, help="Print the logs for the MLflow deployment.")
+def logs_command(kubeconfig: str, notebook_name: str, print_all: bool, mlflow: bool) -> None:
+    """
+    Prints the logs for the specified notebook or DSS component.
+    """
```
Comment (orfeas-k) [inline src/dss/main.py:47]: What about a more verbose doc string? I didn't understand how exactly this worked until I read the whole function including the examples.

**Review by orfeas-k — COMMENTED** [review]
Left some comments and looking into tests now!

File: src/dss/logs.py [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+        logger.info("  dss initialize  to install dss")
+        return
+
+    if not pods:
+        logger.error("Failed to retrieve logs. No pods found.")
+        return
+
+    for pod in pods:
+        # Retrieve logs from the pod
+        logger.info(f"Logs for {pod.metadata.name}:")
```
[the remark this answers is not in the document]
Comment (misohu) [reply src/dss/logs.py:83]: UX team does not recommend using that as they dont operate with pods.

File: src/dss/logs.py (hunk 5 above)
Comment (misohu) [reply src/dss/logs.py:13]: I don't agree ... the argument is None if --mlflow or --all is provided.

File: src/dss/main.py (hunk 8 above)
Comment (misohu) [reply src/dss/main.py:47]: Thats the help message we agreed with UX team in the spec.

**Review by orfeas-k — COMMENTED** [review]
Great test coverage actually, congratz :tada:

File: src/dss/logs.py [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+        name (str): Name of the notebook or deployment to retrieve logs from.
+            This parameter is required for 'notebooks' logs but not for 'mlflow' or 'all'.
+        lightkube_client (Client): The Kubernetes client.
+
+    Returns:
+        None
+    """
+    try:
+        if parts == "notebooks":
+            deployment_notebooks = list(lightkube_client.list(Deployment, namespace=DSS_NAMESPACE))
```
[the remark this answers is not in the document]
Comment (misohu) [reply src/dss/logs.py]: agree

File: src/dss/logs.py [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+            else:
+                logger.error(f"Failed to retrieve logs. Notebook '{name}' does not exist.")
+                logger.info("Run 'dss list' to check all notebooks.")
+                return
+            pods = lightkube_client.list(
+                Pod, namespace=DSS_NAMESPACE, labels=deployment.spec.selector.matchLabels
+            )
+        elif parts == "mlflow":
+            try:
+                deployment_mlflow = lightkube_client.get(
```
[the remark this answers is not in the document]
Comment (misohu) [reply src/dss/logs.py]: agree

File: src/dss/logs.py [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+            for deployment in deployment_notebooks:
+                if deployment.metadata.name == name:
+                    break
+            else:
+                logger.error(f"Failed to retrieve logs. Notebook '{name}' does not exist.")
+                logger.info("Run 'dss list' to check all notebooks.")
+                return
+            pods = lightkube_client.list(
+                Pod, namespace=DSS_NAMESPACE, labels=deployment.spec.selector.matchLabels
+            )
```
[the remark this answers is not in the document]
Comment (misohu) [reply src/dss/logs.py:38]: yes

File: src/dss/utils.py [hunk 13]
```diff
@@ -19,7 +19,7 @@ def wait_for_deployment_ready(
     client: Client,
     namespace: str,
     deployment_name: str,
-    timeout_seconds: int = 60,
+    timeout_seconds: int = 120,
```
[the remark this answers is not in the document]
Comment (misohu) [reply src/dss/utils.py]: Sorry was my local testing ... I removed it :)

File: src/dss/logs.py [hunk 14]
```diff
@@ -0,0 +1,93 @@
...
+        logger.info(f"Logs for {pod.metadata.name}:")
+        try:
+            for line in lightkube_client.log(pod.metadata.name, namespace=DSS_NAMESPACE):
+                # Remove the newline character from the end of the log message
+                line = line.rstrip("\n")
+                logger.info(line)
+        except ApiError as e:
+            logger.error(e)
+            logger.error(
+                f"Failed to retrieve logs. There was a problem while getting the logs for pod {pod.metadata.name}"  # noqa: E501
```
[the remark this answers is not in the document]
Comment (misohu) [reply src/dss/logs.py]: I removed it from here. Let me know if you like it :)

File: src/dss/main.py (hunk 6 above)
Comment (misohu) [reply src/dss/main.py:69]: Added

File: tests/integration/deploy.yaml [hunk 15]
```diff
@@ -0,0 +1,65 @@
+apiVersion: apps/v1
+kind: Deployment
```
[the remark this answers is not in the document]
Comment (misohu) [reply tests/integration/deploy.yaml]: Nice I like it :)

File: tests/integration/test_dss.py [hunk 16]
```diff
@@ -50,7 +58,50 @@ def test_initialize_creates_dss(cleanup_after_initialize) -> None:
...
+    )
+
+    # Check if the command executed successfully
+    assert result.returncode == 0
+
+    # Check if the expected logs are present in the output
+    assert "Starting gunicorn" in result.stderr
+
+
+@pytest.fixture(scope="module")
```
[the remark this answers is not in the document]
Comment (misohu) [reply tests/integration/test_dss.py:104]: Yes it cleans everything even the namespace

File: src/dss/logs.py (hunk 5 above)
Comment (orfeas-k) [reply src/dss/logs.py:13]: yes, but in those cases it's not used at all. The argument always represents a notebook's name.

File: src/dss/main.py (hunk 8 above)
Comment (orfeas-k) [reply src/dss/main.py:47]: Is the doc string also being used in what the user sees?

File: src/dss/main.py (hunk 7 above)
Comment (orfeas-k) [reply src/dss/main.py:66]: What about putting in the doc string like described here? https://click.palletsprojects.com/en/8.1.x/documentation/#preventing-rewrapping

Comment (misohu) [reply src/dss/main.py]: That helped I refactored the help message. Thanks.

File: src/dss/logs.py (hunk 5 above)
Comment (misohu) [reply src/dss/logs.py:13]: It ius actually used in the error message output when there cam be also --mlflow nad --all not just for notebooks :)

**Review by orfeas-k — APPROVED** [review]
Thanks Michal, great work!
